### DISCUSSION TITLE
How to get Request in a route function

### DISCUSSION THREAD

**Discussion by paul-hansen** [rwf2/Rocket#2197] [Questions] [answered]
It seems like being able to access the [Request](https://api.rocket.rs/v0.5-rc/rocket/struct.Request.html) in a route method isn't possible? There used to be a workaround to use FromRequest on a struct that holds a reference to the request but now that doesn't seem to work with the lifetime changes on FromRequest in `0.5-rc`.

I get that it's better to use [FromRequest](https://api.rocket.rs/v0.5-rc/rocket/request/trait.FromRequest.html) in most cases, but it seems limiting to hide the actual request from the route's function. Isn't the whole point of a route to respond to a request?
For example, if the thing I want to do to using the request is very specific to that route and I will never reuse it, why should we have to create a struct that implements a trait just to get access to the fields of Request?

If it isn't possible, is there a benefit to this that I'm not seeing?

Related: Rendering a template to a string https://github.com/SergioBenitez/Rocket/issues/1177#issuecomment-1133189065

**Comment by SergioBenitez** [comment, marked as answer]
See https://rocket.rs/v0.5-rc/guide/faq/#raw-request.

**Reply by paul-hansen** [reply]
Good to see there is a solid answer out there. Sorry for not seeing the FAQ section, I think I looked everywhere else. Hard to find good search results for it since "Request" is used pretty much everywhere and I wasn't sure what to call route "handlers" (e.g. Django calls them "views"). Hopefully having this discussion here will help the search results for the next person.

Not sure I agree with the approach, feels like it will get in the way more than it helps, but I respect that some people may appreciate the hand holding and being forced into making things re-usable. Maybe it will grow on me.

**Reply by SergioBenitez** [reply]
> Not sure I agree with the approach, feels like it will get in the way more than it helps, but I respect that some people may appreciate the hand holding and being forced into making things re-usable. Maybe it will grow on me.

It's not just about making things reusable, but rather about making types mean something. Consider this request handler signature:

```rust
#[get("/")]
fn f(admin: Admin, agent: DesktopAgent, config: &State<Config>) -> Option<File>;
```

What do we know about the handler by reading this signature? We know:

1. It'll only run if an `Admin` user is authenticated and authorized.
2. It'll only run if we can pull a `DesktopAgent` from the request.
3. It fetches a `Config` from managed state (which Rocket guarantees is present at launch-time via [`Sentinel`s](https://api.rocket.rs/v0.5-rc/rocket/trait.Sentinel.html)).
4. It uses this information, and _only_ this information, to fetch a file and returns it if it's present.

That's a lot of information to gleam from just a signature! Consider instead:

```rust
#[get("/")]
fn f(request: &Request) -> Response;
```

What do we know now? Nothing beyond the fact that it's accepting a request and returning a response.

This is at the _heart_ of Rocket's philosophy.

**Reply by paul-hansen** [reply]
Thanks for taking the time to explain!

> 4. It uses this information, and _only_ this information, to fetch a file and returns it if it's present.

You won me over right there 💯. Coming from a Python background I'm all too familiar with the problems when you don't limit scope.

I think my problem was that I was looking for a guard to access the Request, when I should have been looking for a guard to access parts of the request. Which I realized once you explained the reasoning and it seems very obvious to me now.
So for anyone else struggling with this, you can find the guards for accessing parts of a request and more by looking at all of the "Implementations" and "Implementors" here: https://api.rocket.rs/v0.5-rc/rocket/request/trait.FromRequest.html#foreign-impls This makes it so you are very explicit about what parts of the request you are using for each handler. Very neat! 

Thanks for your patience @SergioBenitez, also thanks for jumping on https://github.com/SergioBenitez/Rocket/issues/1177 it's much appreciated! Was not expecting a resolution that fast. ❤️
